lib/browser/mqtt_request_response/subscription_manager.ts

```diff
@@ -0,0 +1,635 @@
+}
+
+export type SubscriptionOrphanedEventListener = (event: SubscriptionOrphanedEvent) => void;
+
+export interface UnsubscribeCompleteEvent {
+    topicFilter: string,
+}
+
+export type UnsubscribeCompleteEventListener = (event: UnsubscribeCompleteEvent) => void;
+
+export enum SubscriptionType {
+    EventStream,
+    RequestResponse,
+}
+
+export interface AcquireSubscriptionConfig {
+    topicFilters: Array<string>,
+    operationId: number,
+    type: SubscriptionType,
+}
+
+export interface ReleaseSubscriptionsConfig {
+    topicFilters: Array<string>,
+    operationId: number,
+}
+
+export enum AcquireSubscriptionResult {
+    Subscribed,
+    Subscribing,
+    Blocked,
+    NoCapacity,
+    Failure,
+}
+
+export interface SubscriptionManagerConfig {
+    maxRequestResponseSubscriptions: number,
+    maxStreamingSubscriptions: number,
+    operationTimeoutInSeconds: number,
+}
+
+enum SubscriptionStatus {
+    Subscribed,
+    NotSubscribed,
+}
+
+enum SubscriptionPendingAction {
+    Nothing,
+    Subscribing,
+    Unsubscribing,
+}
+
+interface SubscriptionRecord {
+    topicFilter: string,
+    listeners: Set<number>,
+    status: SubscriptionStatus,
+    pendingAction: SubscriptionPendingAction,
+    type: SubscriptionType,
+
+    /*
+     * A poisoned record represents a subscription that we will never try to subscribe to because a previous
+     * attempt resulted in a failure that we judge to be "terminal."  Terminal failures include permission failures
+     * and validation failures.  To remove a poisoned record, all listeners must be removed.  For request-response
+     * operations this will happen naturally.  For streaming operations, the operation must be closed by the user (in
+     * response to the user-facing event we emit on the streaming operation when the failure that poisons the
+     * record occurs).
+     */
+    poisoned: boolean,
+}
+
+interface SubscriptionStats {
+    requestResponseCount: number,
+    streamingCount: number,
+    unsubscribingStreamingCount: number,
+}
+
+export class SubscriptionManager extends BufferedEventEmitter {
+    private static logSubject : string = "SubscriptionManager";
+
+    private closed: boolean = false;
+    private records: Map<string, SubscriptionRecord>;
+
+    constructor(private adapter: protocol_adapter.ProtocolClientAdapter, private options: SubscriptionManagerConfig) {
+        super();
+
+        this.records = new Map<string, SubscriptionRecord>();
+
+        this.adapter.addListener(protocol_adapter.ProtocolClientAdapter.SUBSCRIBE_COMPLETION, this.handleSubscribeCompletionEvent.bind(this));
+        this.adapter.addListener(protocol_adapter.ProtocolClientAdapter.UNSUBSCRIBE_COMPLETION, this.handleUnsubscribeCompletionEvent.bind(this));
+        this.adapter.addListener(protocol_adapter.ProtocolClientAdapter.CONNECTION_STATUS, this.handleConnectionStatusEvent.bind(this));
+    }
+
+    close() {
+        this.closed = true;
+
+        this.unsubscribeAll();
+    }
```

Comment:
Not sure if this is an issue or not, but I think, there is a possibility for a subscription to succeed after the `close` is called. Then, that subscription will remain active after the manager is closed.

Here is a hypothetical case:
1. User calls `manager.acquireSubscription("a/b")` --> `this.activateSubscription("a/b")` --> `this.adapter.subscribe("a/b")`
2. User calls `manager.close()` --> `this.unsubscribe("a/b")` --> `this.adapter.unsubscribe("a/b")`
3. MQTT broker processes `SUBSCRIBE` and returns a retryable SUBACK --> `this.activateSubscription("a/b")` --> `this.adapter.subscribe("a/b")`
4. MQTT broker processes `UNSUBSCRIBE`.
5. MQTT broker processes `SUBSCRIBE` with a success.

Even without the `retryable` logic, it's possible that MQTT broker will process subscribe and unsubscribe requests in reversed order:
3. MQTT broker processes `UNSUBSCRIBE`
4. MQTT broker processes `SUBSCRIBE`
But I don't think we can handle this in a robust way.

But the retryable logic probably could be improved by checking `closed` flag.

Reply:
Predicating the resubscribe with a check of this.closed would address, yes?

Beyond that I am unsure how the broker would receive an unsubscribe and a subscribe out of submission order.

Reply:
> Predicating the resubscribe with a check of this.closed would address, yes?

I believe yes, it should help.